### ISSUE TITLE
Is there a way to perform implicit joins using navigation properties (a-la EF)?

### ISSUE THREAD

**Issue by fberasategui** [MicrosoftDocs/powerapps-docs#1054]
Will link [this](https://github.com/MicrosoftDocs/dynamics-365-customer-engagement/issues/1529) issue here, to see if I get better luck

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 89464513-67cd-0821-c0b8-5130bf8b2d16
* Version Independent ID: 7e8041b0-7d9c-4d68-67ef-77b6ec29a146
* Content: [Linq Query Examples (Common Data Service) - PowerApps](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/org-service/linq-query-examples)
* Content Source: [powerapps-docs/developer/common-data-service/org-service/linq-query-examples.md](https://github.com/MicrosoftDocs/powerapps-docs/blob/live/powerapps-docs/developer/common-data-service/org-service/linq-query-examples.md)
* Service: **powerapps**
* GitHub Login: @JimDaly
* Microsoft Alias: **jdaly**

**Comment by JimDaly** [comment]
@fberasategui 
What are you trying to do and what results do you get?
You should try to use one of the lower-level query capabilities (QueryExpression or FetchXml). If you can't get what you want with those, you won't be able to get it with LINQ.

**Comment by fberasategui** [comment]
@JimDaly Okay. First of all thanks for your attention.

I've created a [repro](https://github.com/fberasategui/cds-linq-issue-1054).

As you can see, this is the query I'm trying to perform:

```csharp
var accounts =
       ctx.CreateQuery<Account>()
          .Where(x => x.account_primary_contact.FirstName.StartsWith("John"))
          .ToList();
```

The query itself is pretty simple, in plain english:

> Get me all Accounts whose Primary Contact's First Name Starts with "John"

In Advanced Find (sorry for the spanish here)

![image](https://user-images.githubusercontent.com/13060549/74898286-38a8f800-5378-11ea-8024-ae0cc5fadf3d.png)

Which yields the following `FetchXml`:

```xml
<fetch version="1.0" output-format="xml-platform" mapping="logical" distinct="false">
  <entity name="account">
    <attribute name="name" />
    <attribute name="primarycontactid" />
    <attribute name="telephone1" />
    <attribute name="accountid" />
    <order attribute="name" descending="false" />
    <link-entity name="contact" from="contactid" to="primarycontactid" link-type="inner" alias="ab">
      <filter type="and">
        <condition attribute="firstname" operator="like" value="John%" />
      </filter>
    </link-entity>
  </entity>
</fetch>
```

Which, of course, works just fine.

Now, the LINQ query throws the following exception:

![image](https://user-images.githubusercontent.com/13060549/74898575-fe8c2600-5378-11ea-8f54-299b71e32899.png)

with the following stack trace:

```
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at Microsoft.Xrm.Sdk.IOrganizationService.Execute(OrganizationRequest request)
   at Microsoft.Xrm.Sdk.Client.OrganizationServiceProxy.ExecuteCore(OrganizationRequest request)
   at Microsoft.Xrm.Sdk.Client.OrganizationServiceProxy.Execute(OrganizationRequest request)
   at Microsoft.Xrm.Sdk.Client.OrganizationServiceContext.Execute(OrganizationRequest request)
   at Microsoft.Xrm.Sdk.Linq.QueryProvider.RetrieveEntityCollection(OrganizationRequest request, NavigationSource source)
   at Microsoft.Xrm.Sdk.Linq.QueryProvider.Execute(QueryExpression qe, Boolean throwIfSequenceIsEmpty, Boolean throwIfSequenceNotSingle, Projection projection, NavigationSource source, List`1 linkLookups, String& pagingCookie, Boolean& moreRecords)
   at Microsoft.Xrm.Sdk.Linq.QueryProvider.Execute[TElement](QueryExpression qe, Boolean throwIfSequenceIsEmpty, Boolean throwIfSequenceNotSingle, Projection projection, NavigationSource source, List`1 linkLookups)
   at Microsoft.Xrm.Sdk.Linq.QueryProvider.Execute[TElement](Expression expression)
   at Microsoft.Xrm.Sdk.Linq.QueryProvider.GetEnumerator[TElement](Expression expression)
   at Microsoft.Xrm.Sdk.Linq.Query`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at ConsoleApp6.Program.Main(String[] args) in C:\Users\FedericoBerasategui\source\repos\ConsoleApp6\ConsoleApp6\Program.cs:line 20
```

which suggests me that it's mistakenly picking the "John%" value and using it as a filter value for the `primarycontactid` attribute, instead of using it for the `firstname` attribute as in the FetchXml above. I'll see if I can intercept the exact FetchXml that is getting generated here and come back with that.

Please notice I've provided a clean, as-vanilla-as-possible repro:

  - Using a vanilla `GeneratedCode.cs`, generated using latest CrmSvcUtil 9.1.0.38 from a vanilla D365 instance, as follows:

      `
      CrmSvcUtil.exe /url:https://myorg.crm.dynamics.com/XRMServices/2011/Organization.svc 
      /out:GeneratedCode.cs /username:username /password:password
      `

  -  Using only Nuget Package `Microsoft.CrmSdk.CoreAssemblies`, latest version `9.0.2.23`

## Steps to reproduce:
  - Clone the repo: https://github.com/fberasategui/cds-linq-issue-1054
  - Open `ConsoleApp6.sln` in Visual Studio 2019
  - Edit lines `31-33` of file `Program.cs` providing connection details (OrganizationService Url, username and password)
  - F5

**Comment by fberasategui** [comment]
@JimDaly update:

I've intercepted down the `RetrieveMultipleRequest` that's generated by the LINQ provider:

![image](https://user-images.githubusercontent.com/13060549/74900586-11a1f480-537f-11ea-913c-3d3aa1fb12d4.png)

Notice how there are no `LinkEntities` and it effectively uses the "John%" value for the wrong attribute `primarycontactid`.

This suggests a long-overlooked bug in the LINQ provider.

**Comment by JimDaly** [comment]
@fberasategui 
I appreciate the detailed repro steps.
Typically we don't accept bugs via the documentation feedback process. It is for feedback on the content, not for reporting bugs. Typically, you should open a support request to report issues.
If you report one of the samples published doesn't work - that would be something that should be fixed in the docs. But you are describing a bug.

Although I'm the author of many SDK topics, I'm not the PM owner for all the features. I wrote these samples many years ago.  But considering the effort you provided to supply repro steps,  I'll try to locate who currently owns this part of the SDK and ask them to take a look.

**Comment by JimDaly** [comment]
I've opened a bug for this (including all the details of your repro steps) and assigned it to the team that owns the SDK. 
But since this isn't an actionable doc issue (yet) I'll close this out.
If they report it is by-design, we can identify the limitations and document them.
I've included a link to this issue in the bug.

**Comment by jordimontana82** [comment]
@fberasategui @JimDaly   I think you're using the name of the relationship (account_primary_contact), not the name of the lookup field (primarycontactid), if you use that attribute in the Where clause it should then filter on "primarycontactid" as it does with fetchxml..

**Comment by fowl2** [comment]
@JimDaly did we have an outcome from that bug?

I'm getting similar errors but [LINQ Limitations](https://docs.microsoft.com/en-us/power-apps/developer/data-platform/org-service/use-linq-construct-query#linq-limitations) still doesn't mention anything about navigation properties being unsupported.
